🔥 Overclock edition minting

0xkarmacoma opened the pull request:
See individual commits for details, quick summary of optimizations:
- pack storage so that we can check the edition size + time limit and increment numMinted with 1 SLOAD + 1 SSTORE
- be smart about the order of fields in the packed storage so that we can just increment the whole thing without any bit shuffling
- use internal functions instead of modifiers to avoid spurious SLOADs
- reorder checks in `_isAllowedToMint()` to incur 1 SLOAD vs 3 in most cases
- expose both mint and safeMint, caller can choose if it makes sense to pay for the extra safety

0xkarmacoma committed:
Add new tests for baseline gas snapshots

0xkarmacoma committed:
Pack storage in a single struct

testCreateEditionWithNoRoyalties() (gas: -1930 (-0.663%))
testMintBatch10() (gas: -3585 (-1.275%))
testMintBatch3() (gas: -3718 (-3.575%))
testMintBatch1() (gas: -3756 (-7.058%))
testMintSingle() (gas: -3821 (-7.184%))
testCreateEditionWithEmptyDescription() (gas: -21842 (-7.446%))
testFailMintSingle() (gas: -3821 (-18.416%))
testBurnDecreasesTotalSupply() (gas: -20643 (-20.236%))

0xkarmacoma committed:
Optimize _mintEdition with single SLOAD

That means getting rid of the `notEnded` modified and flattening the checks to make sure we only read from storage once

0xkarmacoma committed:
Declutter _isAllowedToMint()

0xkarmacoma committed:
Optimize _isAllowedToMint() to check for allowlist first

owner mints are the least likely to occur in practice, checking in the most likely order can drastically reduce the amount of storage we need to read

0xkarmacoma committed:
Update snapshot

0xkarmacoma committed:
Optimize _mintEdition()

0xkarmacoma committed:
Expose both mintEdition and safeMintEdition

Customer's choice!

0xkarmacoma commented:
Before:

```
[PASS] testMintSingle() (gas: 57560)
```

After:

```
[PASS] testMintSingle() (gas: 46261)
```

fontanierh commented on contracts/Edition.sol:
```diff
@@ -32,27 +32,29 @@ contract Edition is
-
-    // How many tokens have been burned
-    uint256 public numberBurned;
+    struct EditionState {
```
So the gas savings are because 
- you only need 1 instruction to fetch the whole thing ?
- You picked a specific order for the fields so that you can increment without "bit shuffling" ? Can you explain that one a bit more ? that sounds interesting !

fontanierh replied:
Ok I think I understand by reading the assembly part

0xkarmacoma replied:
the 5 values used to take 5 storage slots but now fit in one so:
- this saves 4 writes on creation
- this saves a bunch of reads on mint
- the increment can be done very minimally as you have seen

fontanierh commented on contracts/Edition.sol:
```diff
@@ -230,43 +230,28 @@ contract Edition is
         pure
     {
-        if (_editionSize > 0 && _numberMinted >= _editionSize) {
+        if (_editionSize > 0 && _numberMinted > _editionSize) {
```
Was it a bug before ?

0xkarmacoma replied:
I think I changed it from "do we still have some room" to "fail if this would go over the limit"

fontanierh commented on contracts/Edition.sol:
```diff
@@ -33,18 +33,20 @@ contract Edition is
         // how many tokens have been burned (can not be more than numberMinted)
         uint56 numberBurned;
+        // Pad to make the state fit in exactly 256 bits
+        uint16 __pad;
```
so it's cheaper to add blank bytes in the struct ? 🤯

axeldelafosse replied:
If I understand correctly, the struct here is 256 bits and we use it to store 5 values that were previously 256 bits each (`uint256`) with smaller value types (`uint16` or `uint56`). The pad is just to add the 16 missing bits, but I'm not sure to understand what would happen if you didn't pad (if the struct was 240 bits)

fontanierh replied:
My guess is that having the struct be exactly 256bit helps for that: https://github.com/showtime-xyz/nft-editions/pull/24/files/d52b79ff41724422840991e976f8e7f91ac7422a..f0d4f6c9c9cfe518c208466b67d25190b739f18a#r1005537187

fontanierh replied:
> we use it to store 5 values that were previously 256 bits each

Indeed ! Didn't notice those integers were previously 256 bits each !

0xkarmacoma replied:
yeah there will be blank bytes anyway because each storage slot is 32 bytes. Adding the padding explicitly lets me control where the blank bytes go

fontanierh commented on contracts/Edition.sol:
```diff
@@ -238,24 +241,36 @@ contract Edition is
+
+            // can not realistically overflow
+            // the fields in EditionState are ordered so that incrementing state increments numberMinted
+            _postState := add(_state, 1)
```
omg 😆

fontanierh reviewed (approved):
pretty crazy

jorgeavaldez commented on contracts/Edition.sol:
```diff
@@ -221,84 +227,143 @@ contract Edition is
         }
 
-        _safeMint(recipient, tokenId);
+        enforceSupplyLimit(_editionSize, _tokenId);
+        enforceTimeLimit(_endOfMintPeriod);
```
i love that these are pure functions now (vs `increaseNumberMinted` that did the validation and update in one function)!

jorgeavaldez commented on contracts/Edition.sol:
```diff
@@ -221,84 +227,143 @@ contract Edition is
+    }
+
+    /// @dev Validates the supply and time limits for minting a single edition with a single SLOAD and SSTORE
+    function _mintPreFlightChecks() internal returns (uint56 _tokenId) {
```
This was surprisingly easy to follow along with considering the dark magic!

jorgeavaldez commented on contracts/Edition.sol:
```diff
@@ -90,26 +94,29 @@ contract Edition is
         }
+
+        state = EditionState({
+            editionSize: uint56(_editionSize),
+            endOfMintPeriod: endOfMintPeriod,
+            royaltyBPS: uint16(_royaltyBPS),
```
i read that this truncates the leading bits (preserving the least significant bits)? is it possible to enforce that the caller pass in uint56? i'm wondering if it would be surprising to pass in (for example) `10000000` and the edition storing something `00`. i'm mostly concerned for people who may stumble upon this and deploy it themselves since it's open source, i think we can be safe about it from the backend :smiley:

0xkarmacoma replied:
yes that's a good point, we could revert if the value is too big to store in a uint56

0xkarmacoma replied:
in [25822d4](https://github.com/showtime-xyz/nft-editions/pull/24/commits/25822d47bdb2ed9bf8c1766f6240313eaa1288ab)

jorgeavaldez reviewed (approved):
:fire: :fire: :fire: 

just one small question about typing in initialize but otherwise lgtm!

0xkarmacoma committed:
Enforce size requirements

We don't want to silently truncate fields like editionSize or endOfMintPeriod

Also:
- rethrow errors in EditionCreator
- more tests

0xkarmacoma committed:
Store salePrice internally in milli-ETH

The external interface does not change, we still expect and return a price in wei. But we store it internally as a uint16 value in milliEth, which has the advantage of natively offering a reasonable range and also replacing the padding in the packed EditionState

Overall behavior should be unchanged, but it's:
- a little cheaper in gas to call setPurchasePrice()
- a little cheaper to call purchase() (still 2 SLOADs but from the same slot instead of 2 different slots)

0xkarmacoma committed:
Use custom error PriceTooLow()

+ more tests

0xkarmacoma committed:
Rename EditionMetadataTest.sol to Edition.t.sol
